index.js

```diff
@@ -94,39 +93,31 @@ module.exports = class Hyperdrive extends ReadyResource {
     return new Hyperdrive(this.corestore, this.key, {
       onwait: this._onwait,
       _checkout: null,
-      _db: this.db,
-      _files: this.files.batch()
+      _db: this.db.batch()
     })
   }
 
-  async flush () {
-    await this.files.flush()
-    return this.close()
+  flush () {
+    return this.db.flush()
   }
 
   async _close () {
-    try {
-      if (this.blobs !== null && (this._checkout === null || this.blobs !== this._checkout.blobs)) {
-        await this.blobs.core.close()
-      }
-      if (!this._batching) await this.db.close()
-      await this.files.close() // workaround to flush the batches for now. TODO: kill the sub!
-    } catch (e) {
-      safetyCatch(e)
+    if (this.blobs && (!this._checkout || this.blobs !== this._checkout.blobs)) {
+      await this.blobs.core.close().catch(safetyCatch)
     }
 
-    if (this._checkout || this._batching) return
-    try {
-      await this.corestore.close()
-    } catch (e) {
-      safetyCatch(e)
+    await this.db.close().catch(safetyCatch)
+
+    if (!this._checkout && !this._batching) {
+      await this.corestore.close().catch(safetyCatch)
     }
   }
 
   async _openBlobsFromHeader (opts) {
     if (this.blobs) return true
 
-    const header = await this.db.getHeader(opts)
+    const db = this.db.tree || this.db
+    const header = await db.getHeader(opts)
```

Comment:
Is that ok? If it's a batch then it will use the `tree` property to get the actual Hyperbee instance

Reply:
Lets add getHeader on the batch instead

Reply:
Addressing this in the other comment: https://github.com/holepunchto/hyperdrive/pull/344#discussion_r1233046936